### PR TITLE
build classmethod

### PR DESCRIPTION
**Pull request by mosquito**
kwargs constructor for create a new url instance.

```python
u = URL.build(
    scheme='http',
    host='127.0.0.1',
    user='foo',
    password='bar',
    port=8000,
    path='/index.html',
    query_string="arg=value1",
    fragment="top"
)

assert str(u) == 'http://foo:bar@127.0.0.1:8000/index.html?arg=value1#top'
```

`scheme` and `host` is required in this case

new version of #57

### CODE REVIEW

**Comment by mosquito** [comment]
@asvetlov I totally rebase and reopen #57. Please review.

File: yarl/__init__.py [hunk 1]
```diff
@@ -192,6 +192,23 @@ def __init__(self, val='', *, encoded=False, strict=False):
...
         self._cache = {}
 
+    @classmethod
+    def build(cls, scheme='', user='', password='', host='', port=None, path='',
+              query_string='', fragment='', *, strict=False, encoded=False):
+
+        assert scheme and host, "You must specify scheme and host"
+
+        return cls(
+            urlunsplit(
```
Comment (asvetlov) [inline yarl/__init__.py]: Implementation looks unnecessary slow and maybe error-prone.

I suggest just passing SplitResult without unsplitting.

Comment (mosquito) [inline yarl/__init__.py]: `SplitResult` has been removed.
`urlunsplit` implementation is so simple and clear (just string concatenation and condition checks)

**Comment by asvetlov** [comment]
No.
Concatenating and parsing result back in `__init__` is very bad and slow idea.

**Comment by mosquito** [comment]
@asvetlov I made tests for it.

First of all I rewrote it like this:
```python
@classmethod
def build(cls, scheme='', user='', password='', host='', port=None, path='',
          query_string='', fragment='', *, strict=False, encoded=False):

    assert scheme and host, "You must specify scheme and host"

    url = cls('%s://%s' % (scheme, host), strict=strict, encoded=encoded)

    if user:
        url = url.with_user(user)

    if password:
        url = url.with_password(password)

    if port:
        url = url.with_port(port)

    if path:
        url = url.with_path(path)

    if query_string:
        if isinstance(query_string, str):
            url = url.with_query(query_string)
        elif isinstance(query_string, dict):
            url = url.with_query(**query_string)
        else:
            url = url.with_query(*query_string)

    if fragment:
        url = url.with_fragment(fragment)

    return url
```

I run ipython and run `%timeit`

```python
In [8]: %timeit URL.build(host="example.net", scheme="http", user="foo", password="bar", fragment="test", query_string="aaa=1")
10000 loops, best of 3: 61.6 µs per loop
```

after that I return `urlunsplit` based implementaion:
```python
@classmethod
def build(cls, scheme='', user='', password='', host='', port=None, path='',
          query_string='', fragment='', *, strict=False, encoded=False):

    assert scheme and host, "You must specify scheme and host"

    return cls(
        urlunsplit((
            scheme, cls._make_netloc(user, password, host, port), path, query_string, fragment
        )),
        strict=strict,
        encoded=encoded
    )
```

ipython's `%timeit` results is:
```python
In [2]: %timeit URL.build(host="example.net", scheme="http", user="foo", password="bar", fragment="test", query_string="aaa=1")
10000 loops, best of 3: 28.7 µs per loop
```

**Comment by mosquito** [comment]
@asvetlov please assist me, what should I do? What changes should be applied?

**Comment by asvetlov** [comment]
Try constructing `SplitResult` as you did in first version and passing this object into `yarl.URL` with `encoded=True`.

**Comment by mosquito** [comment]
@asvetlov you absolutely right. It's better.

```python
In [3]: %timeit yarl.URL.build(scheme="http", host="foo", path="/")
The slowest run took 34.88 times longer than the fastest. This could mean that an intermediate result is being cached.
100000 loops, best of 3: 2.8 µs per loop
```

**Comment by asvetlov** [comment]
Cool!
Please mention new method in `api.rst`

**Comment by mosquito** [comment]
@asvetlov done

File: docs/api.rst [hunk 2]
```diff
@@ -429,7 +450,7 @@ section generates a new *URL* instance.
 
    .. doctest::
 
-      >>> URL('http://example.com/'.with_path('/path/to')
+      >>> URL('http://example.com/').with_path('/path/to')
```
Comment (mosquito) [inline docs/api.rst]: correction for doctests

**Comment by asvetlov** [comment]
There are two questions.
1. Make arguments keyword-only maybe? `URL.build('https', 'guest', 'secret', 'example.com', 1443, '/admin/')` looks messy.
2. `query` parameter might make sense (should be mutually exclusive with `query_string`).

**Comment by mosquito** [comment]
1. Done
2. Done

**Comment by mosquito** [comment]
Was rebased with `master`

**Comment by mosquito** [comment]
@asvetlov any comments?

File: yarl/__init__.py [hunk 3]
```diff
@@ -192,6 +192,45 @@ def __init__(self, val='', *, encoded=False, strict=False):
         self._val = val
         self._cache = {}
 
+    @classmethod
+    def build(cls, *, scheme='', user='', password='', host='', port=None, path='',
+              query=None, query_string='', fragment='', strict=False):
+
+        """ Creates and returns a new URL
+
+        :type scheme: str
```
Comment (asvetlov) [inline yarl/__init__.py]: Pls remove these lines, we don't use sphinx autodoc for the project.

Docstring hould be just 
```
"""Creates and returns a new URL.
"""
```
(no leading space and period at the end).
Just look on other docstrings from the module.

File: yarl/__init__.py [hunk 4]
```diff
@@ -192,6 +192,45 @@ def __init__(self, val='', *, encoded=False, strict=False):
...
+        :type password: str
+        :type host: str
+        :type port: int
+        :type path: str
+        :type query_string: str
+        :type fragment: str
+        :type strict: str
+        """
+        if host or scheme:
+            assert scheme, 'Can\'t build URL with "host" but without "scheme".'
```
Comment (asvetlov) [inline yarl/__init__.py]: Please replace assertions with raising `ValueError`.

File: yarl/__init__.py [hunk 5]
```diff
@@ -192,6 +192,45 @@ def __init__(self, val='', *, encoded=False, strict=False):
...
+            assert host, 'Can\'t build URL with "scheme" but without "host".'
+
+        assert not (query and query_string), "Only one of \"query\" or \"query_string\" should be passed"
```
Comment (asvetlov) [inline yarl/__init__.py]: `assert` -> `ValueError`

**Comment by mosquito** [comment]
@asvetlov can you merge it?

**Comment by fafhrd91** [comment]
I would add quoting

something like https://github.com/fafhrd91/aiohttp/blob/master/aiohttp/http_message.py#L400

**Comment by mosquito** [comment]
@fafhrd91 please add your example as a comment. Your link is obsolete.

**Comment by fafhrd91** [comment]
you need to quote at least path and query string

**Comment by fafhrd91** [comment]
here is proper link https://github.com/aio-libs/aiohttp/blob/master/aiohttp/http_writer.py#L316

**Comment by fafhrd91** [comment]
could you add quoting for path and query, then I will merge

**Comment by mosquito** [comment]
@fafhrd91 done. Actually not needed apply quoting to query string, but if you want I already made it.

**Comment by fafhrd91** [comment]
why do you think we should not quote query string?

**Comment by mosquito** [comment]
@fafhrd91 because I forgot replace `query_sting`. Sorry and thank you.

**Comment by fafhrd91** [comment]
I think `quote` is misleading name

**Comment by mosquito** [comment]
@fafhrd91 should I make any changes?

**Comment by fafhrd91** [comment]
I am not sure. it is in the wild, it would be backward incompatible

**Pull request merged**